vivisect/analysis/i386/golang.py

```diff
@@ -0,0 +1,202 @@
+'''
+Perform further analysis on 32-bit GO language executables.
+
+This module extends code flow analysis to runtime_main for Windows PE binaries.
+GO binaries start from a single export and proceed thru several functions
+that initialize GO.  Specific application code is launched from the GO function
+runtime_main(), which is invoked by "call eax" with an address placed on the
+stack many calls earlier.  Vivisect code flow analysis does not track the
+address; this module finds the address and invokes makeFunction(va).
+
+Samples have been found with different instruction sequences for reaching the
+address of runtime_main(); this module attempts all observed sequences.
+'''
+
+import envi
+import envi.archs.i386.disasm
+
+
+def analyze(vw):
+    '''
+    Perform further analysis on GO language executables.
+    '''
+    # Make sure it is a PE file, with "Go build ID:" in the first few bytes
+    # of the .text segment (versus a .upxN segment of a packed sample).
+    if (vw.getMeta('Format') != 'pe') and \
+       (vw.getMeta('Architecture') != 'i386'):
+        return
+    has_go_build = False
+    for segment in vw.getSegments():
+        va, size, name, filename = segment
+        if name == '.text':
+            bytez = vw.readMemory(va, 10000)
+            k1 = bytez.find(b'Go build ID: ')
+            if k1 != -1:
+                has_go_build = True
+                break
+    if not has_go_build:
+        return
+
+    # There should be just one entry point (the public export).
+    ep = vw.getEntryPoints()
+    if len(ep) != 1:
+        return
+
+    # Search from the entry point function.
+    ep_va = ep[0]
+    bblocks = vw.getFunctionBlocks(ep_va)
+    if not bblocks:
+        return
+    ptr_va, runtime_va = extract_golang_mainmain(vw, bblocks, filename)
+    if runtime_va is None:
+        return
+
+    # Invoke codeflow on runtime_main().
+    vw.makeFunction(runtime_va)
```

Comment:
should we add to the EntryPoints VaSet as well?
@rakuy0 , what do you think?

Reply:
I put this in but it does not cause the pointer to appear in vw.getExports().  Seems like that would be the purpose?

Reply:
thank you.  no, it wouldn't show up in getExports().  getExports() returns a list of symbols which are exported via the file header (ELF/PE).  addEntryPoint() sticks the va into a VaSet named "EntryPoints", which you can access via:

`vw.getVaSetRows('EntryPoints')`